src/ucp/wireup/wireup_cm.c

```diff
@@ -441,19 +447,66 @@ static unsigned ucp_ep_cm_remote_disconnect_progress(ucp_ep_h ucp_ep)
     return 1;
 }
 
+/*
+ * The main thread progress part of connection establishment on server side
+ */
+static unsigned ucp_cm_server_connect_progress(void *arg)
+{
+    ucp_ep_h ucp_ep = arg;
+
+    UCS_ASYNC_BLOCK(&ucp_ep->worker->async);
+    ucs_debug("ep %p: CM server connect progress, flags 0x%x", ucp_ep,
+              ucp_ep->flags);
+    ucp_wireup_remote_connected(ucp_ep);
+    UCS_ASYNC_UNBLOCK(&ucp_ep->worker->async);
+    return 1;
+}
+
+static int ucp_ep_conn_cb_cmp(const ucs_callbackq_elem_t *elem, void *arg)
+{
+    ucp_ep_h ep = arg;
+    ucp_cm_client_connect_progress_arg_t *client_progress_arg;
+
+    if (elem->cb == ucp_cm_client_connect_progress) {
+        client_progress_arg = elem->arg;
+        return client_progress_arg->ucp_ep == ep;
+    } else if (elem->cb == ucp_cm_server_connect_progress) {
+        return ep == elem->arg;
+    } else {
+        return 0;
+    }
+}
+
 static unsigned ucp_ep_cm_disconnect_progress(void *arg)
 {
     ucp_cm_disconnect_progress_arg_t *progress_arg = arg;
     ucp_ep_h ucp_ep                                = progress_arg->ucp_ep;
     uct_ep_h uct_cm_ep                             = progress_arg->uct_cm_ep;
     ucs_async_context_t *async                     = &ucp_ep->worker->async;
+    uct_worker_cb_id_t prog_id;
     ucp_request_t *close_req;
 
     UCS_ASYNC_BLOCK(async);
 
-    ucs_trace("ep %p: got remote disconnect, cm_ep %p", ucp_ep, uct_cm_ep);
+    ucs_debug("ep %p: got remote disconnect, cm_ep %p", ucp_ep, uct_cm_ep);
     ucs_assert(ucp_ep_get_cm_uct_ep(ucp_ep) == uct_cm_ep);
 
+    /* due to progress queue does not guarantee ordering, we can't progress
+     * disconnect prior connect, reschedule */
+    if (ucs_callbackq_find_if(&ucp_ep->worker->uct->progress_q,
```

Comment:
can we use ep state flags instead of this?

Reply:
Then we need extra flags to indicate that async CBs were invoked and handlers are scheduled. We don't have much space in the EP.